src/ray/core_worker/core_worker.cc

```diff
@@ -12,23 +12,37 @@ CoreWorker::CoreWorker(const enum WorkerType worker_type,
       store_socket_(store_socket),
       raylet_socket_(raylet_socket),
       worker_context_(worker_type, job_id),
-      raylet_client_(raylet_socket_,
-                     ClientID::FromBinary(worker_context_.GetWorkerID().Binary()),
-                     (worker_type_ == ray::WorkerType::WORKER),
-                     worker_context_.GetCurrentJobID(), ToTaskLanguage(language_)),
       task_interface_(*this),
       object_interface_(*this),
       task_execution_interface_(*this) {
-  // TODO(zhijunfu): currently RayletClient would crash in its constructor if it cannot
-  // connect to Raylet after a number of retries, this needs to be changed
-  // so that the worker (java/python .etc) can retrieve and handle the error
-  // instead of crashing.
   auto status = store_client_.Connect(store_socket_);
   if (!status.ok()) {
     RAY_LOG(ERROR) << "Connecting plasma store failed when trying to construct"
                    << " core worker: " << status.message();
     throw std::runtime_error(status.message());
   }
+
+  // NOTE(zhijunfu): For non-driver worker, the initialization of
+  // raylet client is delayed to when `TaskExecutionInterface::Run()`
+  // is called, as it is until that time the port for the worker
+  // rpc server can be determined, and this information needs to be
+  // included when worker registers to raylet.
+  if (worker_type_ == WorkerType::DRIVER) {
```

Comment:
I think it's a bit ugly to have these checks for whether a worker is a driver or not. The same goes for the runtime check in `CoreWorkerTaskExecutionInterface::Run` that this is a driver. Can we try to split workers and drivers into separate classes, or is there a reason why we shouldn't do that?

Also, I think it would be fine to have the driver also start an RPC server, as long as its interface is limited to a subset of a normal worker's.

Reply:
Good point. Agreed that it would be cleaner to divide it into `driver` and `worker`, thus we could remove the checks, and also `driver` doesn't actually have a `TaskExecutionInterface`.

For the RPC server, I think driver currently doesn't need one as it doesn't accept tasks from raylet or other workers/drivers, it would need an IO thread though to handle GCS notifications in order to know the location for a given actor.

Reply:
So is the plan to split the `driver` and `worker` in a separate PR?

Reply:
I think it can be done in this PR, as otherwise this PR would still need to depend on that one. But I don't feel strong on this, if you prefer to use a separate PR, I think it's also fine.